lean4-infoview/src/infoview/event.ts

```diff
@@ -5,20 +5,39 @@ import type { Disposable } from 'vscode-languageserver-protocol';
  * to all registered handlers. Handlers can be registered using `on`. */
 export class EventEmitter<E> {
     private handlers: ((_: E) => void)[] = [];
+    private handlersWithKey = new Map<string, ((_: E) => void)[]>();
```

Comment:
What is the purpose of keys? Instead of having keys, one can alternatively fire all the handlers for a given event and check whether the key-parameter matches the expected one in specific handlers, right? In this case, something like
```typescript
  useEvent(ec.events.goToDefinition, key => { if (key === interactiveCodeTagId) { execGoToLoc() } })
```
since, if I understand correctly, this doesn't add any new functionality, is it an optimization? Have you measured it to have an observable impact? If not, I would just do the check in the handler in order to have less code.

Reply:
I haven't measured it, but the main issue I see with not using a key parameter is that [React will run both the code that adds the handler as well as the code that removes it on every re-render](https://react.dev/reference/react/useEffect#connecting-to-an-external-system).

Since removing the handler without the key is linear in the amount of handlers and an amount of handlers that is linear in the size of the tree is being removed, removing all handlers on every re-render is quadratic in the size of the tree (just the delaboration of which is already causing performance issues for some users).

This seems like a potential performance bug to me.

Reply:
I see what you mean! A couple of comments:
- The event handler will not be re-added on every render, but rather only on those renders on which the dependencies have changed. I defined `deps` in all occurrences of `useEvent` now.
- There is still an issue you pointed out in the linear complexity of `filter` invoked during handler removal; I changed it to use handler IDs instead, which should be amortized constant time.
- The issue of linear complexity remains in the case of an event firing, at which point all the keyless handlers would be invoked; so let's keep keyed handlers.